Start to handle performImmediateActionHitTestAtLocation for remote frames

pascoej opened the pull request:
#### 96882cf2b4901d7c7d158c52ca645e28b6545c3d
<pre>
Start to handle performImmediateActionHitTestAtLocation for remote frames
<a href="https://bugs.webkit.org/show_bug.cgi?id=273330">https://bugs.webkit.org/show_bug.cgi?id=273330</a>
<a href="https://rdar.apple.com/problem/127122951">rdar://problem/127122951</a>

Reviewed by Charlie Wolfe.

This change starts to tell the UI process to continue hit testing whenever it
encounter a RemoteFrame.

It does not yet translate the coordinates returned in WebHitTestResultData to be
relative to the main frame. I plan to do that in <a href="https://rdar.apple.com/127120734">rdar://127120734</a>.

* Source/WebCore/rendering/HitTestResult.h:
* Source/WebKit/Shared/WebHitTestResultData.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::performImmediateActionHitTestAtLocation):
(WebKit::WebPageProxy::didPerformImmediateActionHitTest):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WKImmediateActionController.mm:
(-[WKImmediateActionController immediateActionRecognizerWillPrepare:]):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::performImmediateActionHitTestAtLocation):
(WebKit::WebPage::lookupTextAtLocation):

Canonical link: <a href="https://commits.webkit.org/278067@main">https://commits.webkit.org/278067@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de1ce241da1d40fa2519037b0dfe39ceae3cbb0c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49377 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28659 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52410 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/52616 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/50 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51681 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34674 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26279 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/52616 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51477 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26206 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42530 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/52616 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23661 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/43713 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7745 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45573 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44218 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54128 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24458 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/20644 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/47663 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25730 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/42737 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/46656 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10854 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26569 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25454 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -11838,6 +11838,10 @@ void WebPageProxy::immediateActionDidComplete()
 {
+    if (result.remoteUserInputEventData) {
+        // rdar://127120734 Translate coordinates
+        performImmediateActionHitTestAtLocation(result.remoteUserInputEventData->targetFrameID, result.remoteUserInputEventData->transformedPoint);
```
Shouldn’t we return here? Otherwise `didPerformImmediateActionHitTest()` might be called more times than expected.

SupaYo replied:
[]()

charliewolfe commented on Source/WebKit/UIProcess/WebPageProxy.h:
```diff
@@ -1791,7 +1791,7 @@ class WebPageProxy final : public API::ObjectImpl<API::Object::Type::Page>, publ
 #if PLATFORM(MAC)
     API::HitTestResult* lastMouseMoveHitTestResult() const { return m_lastMouseMoveHitTestResult.get(); }
-    void performImmediateActionHitTestAtLocation(WebCore::FloatPoint);
+    void performImmediateActionHitTestAtLocation(const WebCore::FrameIdentifier&, WebCore::FloatPoint);
```
nit - We usually just pass FrameIdentifier by value

charliewolfe commented on Source/WebKit/UIProcess/mac/WKImmediateActionController.mm:
```diff
@@ -175,7 +175,7 @@ - (void)immediateActionRecognizerWillPrepare:(NSImmediateActionGestureRecognizer
     immediateActionRecognizer.animationController = nil;
 
-    RefPtr { _page.get() }->performImmediateActionHitTestAtLocation([immediateActionRecognizer locationInView:immediateActionRecognizer.view]);
+    RefPtr { _page.get() }->performImmediateActionHitTestAtLocation(RefPtr { _page.get() }->mainFrame()->frameID(), [immediateActionRecognizer locationInView:immediateActionRecognizer.view]);
```
`_page` doesn't seem like it needs to be protected again.

charliewolfe commented on Source/WebKit/Shared/WebHitTestResultData.h:
```diff
@@ -89,6 +90,7 @@ struct WebHitTestResultData {
     enum class ElementType : uint8_t { None, Audio, Video };
     ElementType elementType;
     std::optional<FrameInfoData> frameInfo;
+    std::optional<WebCore::RemoteUserInputEventData> remoteUserInputEventData;
```
Is this actually being used right now? If so, don't you need to change `WebHitTestResultData.serialization.in`?

charliewolfe commented on Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
```diff
@@ -818,37 +820,51 @@ static void drawPDFPage(PDFDocument *pdfDocument, CFIndex pageIndex, CGContextRe
+    RefPtr currentFrame = WebProcess::singleton().webFrame(frameID);
+    if (!currentFrame)
         return;
+    RefPtr localCurrentFrame = currentFrame->protectedCoreLocalFrame();
```
nit - this causes unnecessary reference churn, can just be:
 ```suggestion
     RefPtr localCurrentFrame = currentFrame->coreLocalFrame();
```

pascoej committed:
Start to handle performImmediateActionHitTestAtLocation for remote frames
https://bugs.webkit.org/show_bug.cgi?id=273330
rdar://problem/127122951

Reviewed by Charlie Wolfe.

This change starts to tell the UI process to continue hit testing whenever it
encounter a RemoteFrame.

It does not yet translate the coordinates returned in WebHitTestResultData to be
relative to the main frame. I plan to do that in rdar://127120734.

* Source/WebCore/rendering/HitTestResult.h:
* Source/WebKit/Shared/WebHitTestResultData.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::performImmediateActionHitTestAtLocation):
(WebKit::WebPageProxy::didPerformImmediateActionHitTest):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/WKImmediateActionController.mm:
(-[WKImmediateActionController immediateActionRecognizerWillPrepare:]):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/mac/WebPageMac.mm:
(WebKit::WebPage::performImmediateActionHitTestAtLocation):
(WebKit::WebPage::lookupTextAtLocation):

Canonical link: https://commits.webkit.org/278067@main
